File: cpp/3210_number_of_special_chars/number_of_special_chars.cpp
#include <cassert>
#include <string>
#include <unordered_set>
#include <vector>
using namespace std;

class Solution {
 public:
  int numberOfSpecialChars(string word) {
    unordered_set<char> s(word.begin(), word.end());
    int ans = 0;
    for (char c = 'a'; c <= 'z'; c++) {
      if (s.count(c) && s.count(c - 'a' + 'A')) {
        ans++;
      }
    }
    return ans;
  }
};

int main() {
  vector<tuple<string, int>> tests{
      {"aaAbcBC", 3},
      {"abc", 0},
      {"abBCab", 1},
  };

  for (auto [word, expected] : tests) {
    assert(Solution().numberOfSpecialChars(word) == expected);
  }
}

File: golang/3121_number_of_special_chars/number_of_special_chars.go
// Package main ...
package main

func numberOfSpecialChars(word string) int {
	lastLow := [26]int{}
	firstUp := [26]int{}
	for i := range lastLow {
		lastLow[i] = -1
		firstUp[i] = -1
	}
	for i, c := range word {
		if c >= 'a' && c <= 'z' {
			lastLow[c-'a'] = i
		} else {
			if firstUp[c-'A'] == -1 {
				firstUp[c-'A'] = i
			}
		}
	}
	ans := 0
	for i := 0; i < 26; i++ {
		if lastLow[i] != -1 && firstUp[i] != -1 && lastLow[i] < firstUp[i] {
			ans++
		}
	}
	return ans
}

func main() {
	tests := []struct {
		word string
		ans  int
	}{
		{"aaAbcBC", 3},
		{"abc", 0},
		{"AbBCab", 0},
	}

	for _, test := range tests {
		if got := numberOfSpecialChars(test.word); got != test.ans {
			panic(test)
		}
	}
}
